### PR TITLE
New reader for Himawari L2 NOAA enterprise cloud products.

### PR DESCRIPTION
**Pull request by simonrp84**
This PR adds a reader for the L2 cloud products products by NOAA for the Himawari satellites. These products are distributed via NOAA's big data programme in netCDF4 format and can be found at, for example: [AHI-L2-FLDK-Clouds](https://noaa-himawari8.s3.amazonaws.com/AHI-L2-FLDK-Clouds/).

 - [x] Add support for all cloud mask datasets
 - [x] Add support for all cloud phase datasets
 - [x] Add support for all cloud height datasets
 - [x] Tests added
 - [x] Fully documented

### CODE REVIEW

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by djhoese** [comment]
How do these differ from ABI L2? Should we share the python code?

**Comment by simonrp84** [comment]
I think the ABI ones contain enough metadata to create an area Def while the AHI ones do not (just lat lon values). Aside from that, they're broadly similar.

**Comment by djhoese** [comment]
Instead of defining every possible L2 product in the YAML, could we dynamically load these and use the variable name from the file?

File: satpy/readers/ahi_l2_nc.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        # Basic check to ensure we're processing a full disk (2km) scene.n
+        if self.nlines != 5500 or self.ncols != 5500:
+            raise ValueError("Input L2 file is not a full disk Himawari scene. Only full disk data is supported.")
+
+        pdict = {'cfac': 20466275, 'lfac': 20466275, 'coff': 2750.5, 'loff': 2750.5, 'a': 6378137.0, 'h': 35785863.0,
+                 'b': 6356752.3, 'ssp_lon': 140.7, 'nlines': self.nlines, 'ncols': self.ncols, 'scandir': 'N2S'}
+
+        aex = get_area_extent(pdict)
+
+        pdict['a_name'] = 'Himawari_Area'
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/readers/ahi_l2_nc.py:127]: No, not really. The `self.nc.attrs['cdm_data_type']` attrs stores an area name but it's always `Full Disk`, which isn't very meaningful. I can set `pdict['a_name']` to that if you prefer though.

**Comment by simonrp84** [comment]
> Instead of defining every possible L2 product in the YAML, could we dynamically load these and use the variable name from the file?

We could but I quite like having them explicitly defined as it makes it easier to browse what products can be loaded before you even download the files. If you have strong feelings about it then I can change though.

**Comment by djhoese** [comment]
With other L2/EDR readers we (I) have moved to dynamic loading. I agree it is nice to be able to look at the YAML and say "yep, it's called that and it's supported", but I would argue that you can do the same with dynamic variables by looking at the YAML file patterns "yep, this file pattern is listed and the variable I want is 2D, so it must be supported".

I'd rather not have to maintain a potentially long list of variables in the YAML and update every time a new L2 product comes out. Additionally, for polar-orbiters the Longitude/Latitude handling gets really sketchy (not that that's an issue for this reader).

I can be convinced otherwise...but not as easily as I have been on other issues since I'd likely be the person to add new products.

**Comment by simonrp84** [comment]
Can you point me at an example of a dynamically loaded YAML? Happy to change, but am unfamiliar with the syntax for this...

**Comment by djhoese** [comment]
The YAML isn't dynamic, the python code is. Any reader that defines the `available_datasets` method is doing this:

https://github.com/pytroll/satpy/blob/c2ed74b139acc28a02f8b859828dfacc0f45cf6f/satpy/readers/file_handlers.py#L204

That includes my recently merged (and updated version of your) VIIRS EDR reader. That one got a little complicated due to the different versions of lon/lat variables in each file.

**Comment by djhoese** [comment]
Regardless of your feelings about the dynamic dataset definitions, the tests are failing and will need to be fixed before merging.

File: satpy/readers/ahi_l2_nc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/ahi_l2_nc.py:134]: If there's not an in-file attribute for this, then can we guess a decent abbreviation/region name based on the filename?

**Comment by simonrp84** [comment]
Looks like the only remaining failing test is for the CF writer code rather than anything this PR has touched.

File: satpy/readers/ahi_l2_nc.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Satpy developers
```
Comment (mraspaud) [inline satpy/readers/ahi_l2_nc.py:3]:
```suggestion
# Copyright (c) 2023 Satpy developers
```

File: satpy/readers/ahi_l2_nc.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Reader for Himawari L2 cloud products from NOAA's big data programme.
```
Comment (mraspaud) [inline satpy/readers/ahi_l2_nc.py:18]: Would this be the right place to give a link to the format documentation?

File: satpy/readers/ahi_l2_nc.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+EXPECTED_DATA_AREA = 'Full Disk'
+
+
+class HIML2NCFileHandler(BaseFileHandler):
+    """File handler for Himawari L2 NOAA enterprise data in netCDF format."""
+
+    def __init__(self, filename, filename_info, filetype_info):
+        """Initialize the reader."""
+        super(HIML2NCFileHandler, self).__init__(filename, filename_info,
```
Comment (mraspaud) [inline satpy/readers/ahi_l2_nc.py:66]:
```suggestion
        super().__init__(filename, filename_info,
```

**Review by mraspaud — COMMENTED** [review]
Thanks for the reader! Just a few comments inline, but looks good otherwise.

**Comment by simonrp84** [comment]
Have adopted all your suggestions, @mraspaud!

There was previously a test to ensure that the warning was raised about the full disk area. As this is now an info message I'm not testing it but can add a test if it really needs one..

**Review by mraspaud — APPROVED** [review]
LGTM!

**Pull request merged**